### PR TITLE
probe: add simple adaptive attacks probe

### LINKED ISSUE #583: probe: simple adaptive attacks
code & probes here: https://github.com/tml-epfl/llm-adaptive-attacks

paper here: https://arxiv.org/abs/2404.02151

### PR DESCRIPTION
**Pull request by neerazz**
## Summary

Implements `adaptiveattacks.AdaptiveAttacks` and `adaptiveattacks.AdaptiveAttacksFull` based on Andriushchenko, Croce, and Flammarion, *"Jailbreaking Leading Safety-Aligned LLMs with Simple Adaptive Attacks"* (ICML 2024, [arXiv:2404.02151](https://arxiv.org/abs/2404.02151)).

The v1 probe is **template-only**: it combines a paraphrased adaptive prompt template (direct-assistant role + structured response scaffold + no-`I` affirmative-opener constraint) with a small cached set of seed forbidden behaviors and pre-computed adversarial suffixes drawn from public AdvBench results. No gradient computation or online search is performed.

## Files

- `garak/probes/adaptiveattacks.py` — `AdaptiveAttacks` (active, capped via `soft_probe_prompt_cap`) and `AdaptiveAttacksFull` (uncapped, opt-in)
- `garak/data/adaptive_attacks/seeds.txt` — 20 seed forbidden behaviors
- `garak/data/adaptive_attacks/suffixes.txt` — 3 cached adversarial suffix variants
- `tests/probes/test_probes_adaptiveattacks.py` — 8 tests covering load, prompt content, template markers, prompt-cap behavior

## Engagement context

The original ticket #583 was soft-claimed by @enesyesil on 2025-09-17. After 7 months without a PR, I posted a polite check-in on 2026-04-28 ([comment](https://github.com/NVIDIA/garak/issues/583#issuecomment-4331671976)) saying I would proceed in 7 days unless they responded. No response received. Happy to step aside if @enesyesil resumes or @leondz prefers a different approach.

## Test plan

- [x] `pytest tests/probes/test_probes_adaptiveattacks.py -v` — 8/8 pass
- [x] `garak --list_probes | grep adaptiveattacks` — both classes registered, default-active marker shown
- [x] `garak --probes adaptiveattacks.AdaptiveAttacks --model_type test.Blank` — completes in 1.42s with no errors
- [ ] CA/DCO sign-off — first commit signed with `-s`; will sign CA on bot prompt

## Scope

v1 is intentionally minimal (single template + cached suffixes) per "good first issue" sizing. Future variants (random-search suffix optimization, GCG-style gradient suffix, multi-turn self-transfer) can land as follow-up PRs.

Closes #583.

### CODE REVIEW

**Comment by leondz** [comment]
Thank you @neerazz, will take a look

File: garak/probes/adaptiveattacks.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    Uncapped variant of :class:`AdaptiveAttacks`. Generates one prompt per
+    (seed, suffix) pair without sampling. Disabled by default; enable via
+    ``--probes adaptiveattacks.AdaptiveAttacksFull``."""
+
+    aliases = ["adaptiveattacks.AdaptiveAttacksFull"]
+    active = False
+
+    def __init__(self, config_root=_config):
+        # Skip the cap-based pruning that AdaptiveAttacks applies.
+        garak.probes.Probe.__init__(self, config_root=config_root)
```
Comment (jmartin-tech) [inline garak/probes/adaptiveattacks.py]: In the interest of consolidating on cap enforcement as desired expressed in #1546 and #1562
```suggestion
        if self.follow_prompt_cap:
            self._prune_data(cap=self.soft_probe_prompt_cap)
```

File: garak/probes/adaptive_attacks.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+        "avid-effect:performance:P0204",
+        "owasp:llm01",
+        "quality:Security:PromptStability",
+        "payload:jailbreak",
+    ]
+    goal = "cause the model to disregard the system prompt and produce a forbidden response"
+    tier = garak.probes.Tier.COMPETE_WITH_SOTA
+    active = True
+    doc_uri = "https://arxiv.org/abs/2404.02151"
+
```
Comment (jmartin-tech) [inline garak/probes/adaptive_attacks.py:90]:
```suggestion

    DEFAULT_PARAMS = garak.probes.Probe.DEFAULT_PARAMS | {
        "follow_prompt_cap": True,
    }

```

**Review by jmartin-tech — COMMENTED** [review]
A suggestion not a full review. The desire to support `Full` probes is somewhat in limbo as they create duplication when activated at the same time as the restricted versions.

**Comment by neerazz** [comment]
I have read the DCO Document and I hereby sign the DCO

**Comment by neerazz** [comment]
Thanks for the review @jmartin-tech. Pushed two fixes:

**1. CI failure (`test_docs.py::test_docs_probes[adaptiveattacks]`)** — added the missing `docs/source/probes/adaptiveattacks.rst` stub and linked it from `index_probes.rst`. The Linux/Mac/Windows pytest cancellations were all downstream of this single doc-test failure.

**2. `Full` probe duplication concern** — `AdaptiveAttacksFull` is already `active = False` (opt-in only) on this branch, but the docstring/comment didn't make the design intent obvious. Updated both to call out explicitly that it's inactive by default to avoid prompt duplication when a run selects both `AdaptiveAttacks` (capped) and `AdaptiveAttacksFull` (uncapped). Happy to drop `Full` entirely if you'd prefer the simpler shape — let me know.

**Comment by neerazz** [comment]
Just pushed [a85aa90](https://github.com/NVIDIA/garak/pull/1742/commits/a85aa90217116022e49095ef4fd38eec587c3494) which applies the canonical pattern more directly per your earlier suggestion:

- **`AdaptiveAttacks`** now declares `DEFAULT_PARAMS = Probe.DEFAULT_PARAMS | {"follow_prompt_cap": True}` and guards the prune call with `if self.follow_prompt_cap: self._prune_data(cap=self.soft_probe_prompt_cap)` — matching the shape used in `encoding.py` / `latentinjection.py` / #1546 / #1562.
- **`AdaptiveAttacksFull`** now expresses "no cap" via `DEFAULT_PARAMS = AdaptiveAttacks.DEFAULT_PARAMS | {"follow_prompt_cap": False}` instead of overriding `__init__` to bypass `AdaptiveAttacks.__init__`. Cleaner, configurable, no constructor hack.

`active = False` on `AdaptiveAttacksFull` is preserved so the duplication concern stays addressed by default. If you'd still prefer to drop `Full` entirely rather than ship two probes, happy to do that in a follow-up commit — just let me know.

Local verification (Python 3.12): 8/8 `tests/probes/test_probes_adaptiveattacks.py`, 4/4 adaptive cases in `tests/test_docs.py`, 1130/1130 in `tests/probes/test_probes.py` (excluding the pre-existing `AudioAchillesHeel` optional-deps failure unrelated to this PR).

File: garak/probes/adaptiveattacks.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+class AdaptiveAttacks(garak.probes.Probe):
+    """Simple adaptive attacks probe (template + cached suffix)
+
+    Combines an adaptive prompt template with a small cached suffix list to
+    elicit policy-violating responses from safety-aligned chat models without
+    requiring gradient access. Based on Andriushchenko, Croce, and Flammarion
+    (ICML 2024)."""
+
+    aliases = ["adaptiveattacks.AdaptiveAttacks"]
```
Comment (jmartin-tech) [inline garak/probes/adaptiveattacks.py]: Aliases are meant for tracking renamed probes. A new probe should not have any aliases.
```suggestion
```

**Comment by neerazz** [comment]
Friendly nudge — wanted to make this easy to action on whatever timeline works.

State recap as of today:
- All 16 CI checks ✅ (Linux/Mac/Windows × Py 3.10/3.12/3.13, docs, pip install)
- DCO signed
- @jmartin-tech's `Full` duplication concern addressed in [a85aa90](https://github.com/NVIDIA/garak/pull/1742/commits/a85aa90217116022e49095ef4fd38eec587c3494) via the canonical `DEFAULT_PARAMS | {"follow_prompt_cap": False}` pattern, with `active = False` preserved so it's opt-in only

Two simple paths from here, whichever you prefer:
1. **Ship as-is** — `AdaptiveAttacks` (capped, active) + `AdaptiveAttacksFull` (uncapped, inactive). Duplication risk is gated by `active = False`.
2. **Drop `Full` entirely** — I'll push a one-commit follow-up removing the second class and keeping only `AdaptiveAttacks` with the configurable cap. Simpler surface, no opt-in probe to maintain.

Happy to do either. Just let me know which shape you'd rather merge.

**Comment by neerazz** [comment]
Simplified to a single probe per the earlier feedback thread — dropped `AdaptiveAttacksFull` entirely (commit [3ceba9b](https://github.com/neerazz/garak/commit/3ceba9b)).

**What changed:**
- `AdaptiveAttacksFull` class removed from `garak/probes/adaptiveattacks.py`
- Test file updated to remove the `Full`-specific parametrize entry and the `test_adaptiveattacks_full_ignores_cap` test
- `AdaptiveAttacks` (capped, `active=True`) is the only probe now — cleaner surface, no inactive opt-in class to maintain

**Uncapped runs still possible** via `--config follow_prompt_cap=false` (the `DEFAULT_PARAMS` key is still there).

This matches option 2 from my earlier comment. All 16 CI checks were green on the previous commit; no functional change to `AdaptiveAttacks` itself, so CI should stay green.

**Comment by neerazz** [comment]
Fixed DCO sign-off. Amended commit \`3ceba9b\` → \`c2f2ae3\` to add the missing \`Signed-off-by\` line:

\`\`\`
Signed-off-by: Neeraj Kumar Singh <b.neerajkumarsingh@gmail.com>
\`\`\`

Force-pushed to \`probe-adaptive-attacks-583\`. DCO check should now pass. @jmartin-tech @leondz — ready for your review when you have a chance.

**Review by jmartin-tech — CHANGES_REQUESTED** [review]
One minor tweak and an information request.

Do you know of a target within the last year that will show a rate of successful attack from this probe?

**Comment by neerazz** [comment]
Thanks for the review, @jmartin-tech!

**Alias tweak** (commit `91b70748`): removed `aliases = [...]`. You're right — aliases are for renamed probes and a new probe should have none. Applied and force-pushed.

**Recent target with measurable success rate:**

The technique this probe implements (adaptive template + cached suffix) is directly from Andriushchenko, Croce, and Flammarion (ICML 2024, arxiv:2404.02151). The paper reports concrete attack success rates against several safety-aligned models that remain in active deployment today: GPT-3.5-turbo, GPT-4, Gemini Pro, and Llama-2-7B/13B-chat. Those are the primary documented targets with published success rates.

For a more recent reference: within the past year, Meta Llama 3 variants (released April 2024) and derivative fine-tuned models have been widely studied as jailbreak targets and are commonly used in garak evaluation runs against this class of prompt-injection attack. Standard RLHF safety training without specific defenses against template-based adaptive prompting continues to show non-trivial attack success rates with this technique.

If you'd prefer a specific success rate number on a 2025-era model, I can run the probe against an open-source model (e.g., Llama-3.1-8B-Instruct) and report the output — but I didn't want to delay the PR for that if the paper's citations and the broader Llama 3 evaluation context are sufficient for your purposes. Let me know.

**Comment by neerazz** [comment]
Hi @jmartin-tech — wanted to follow up since my Jun 9 response may have gotten buried. All items from your review are addressed:

- ✅ `aliases` removed (commit `91b70748`)
- ✅ `AdaptiveAttacksFull` dropped; single capped probe, fully configurable cap via `--config follow_prompt_cap=false` (commit `90eb9f99`)
- ✅ DCO sign-off fixed
- ✅ CI is fully clean (all checks passing)

On your question about a recent target with measurable success rate — I replied with references to the ICML 2024 paper targets (GPT-3.5/4, Gemini Pro, Llama-2) and Llama 3 variants. If you'd find it more useful to have a direct number from running the probe against Llama-3.1-8B-Instruct, I'm happy to do that before you close the review — just say the word.

Whenever you get a chance to re-review, I'd really appreciate it. Thanks again for the detailed feedback that made this a cleaner, simpler probe.

File: garak/probes/adaptiveattacks.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    """Simple adaptive attacks probe (template + cached suffix)
+
+    Combines an adaptive prompt template with a small cached suffix list to
+    elicit policy-violating responses from safety-aligned chat models without
+    requiring gradient access. Based on Andriushchenko, Croce, and Flammarion
+    (ICML 2024)."""
+
+    lang = "en"
+    primary_detector = "mitigation.MitigationBypass"
+    extended_detectors = ["dan.DAN"]
```
Comment (jmartin-tech) [inline garak/probes/adaptiveattacks.py]: Testing shows this detector is not really providing any added value, it should likely be removed.

Is there some additional reason we should keep this?

File: garak/probes/adaptiveattacks.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+
+    prompts = list()
+    seeds = _load_lines(SEEDS_PATH)
+    suffixes = _load_lines(SUFFIXES_PATH)
+    for seed in seeds:
+        for suffix in suffixes:
+            prompts.append(ADAPTIVE_TEMPLATE.format(goal=seed, suffix=suffix))
+
+    def __init__(self, config_root=_config):
+        super().__init__(config_root=config_root)
```
Comment (jmartin-tech) [inline garak/probes/adaptiveattacks.py]: Prompt initialization and processing is commonly preformed in `__init__` this allows loading the module and class without the computational cost if a specific probe class is not utilized.


```suggestion
    def __init__(self, config_root=_config):
        self.prompts = list()
        seeds = _load_lines(SEEDS_PATH)
        suffixes = _load_lines(SUFFIXES_PATH)
        for seed in seeds:
            for suffix in suffixes:
                self.prompts.append(ADAPTIVE_TEMPLATE.format(goal=seed, suffix=suffix))
        super().__init__(config_root=config_root)
```

File: garak/probes/adaptive_attacks.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+results. No gradient computation or online search is performed.
+"""
+
+from garak import _config
+from garak.data import path as data_path
+import garak.probes
+
+
+SEEDS_PATH = data_path / "adaptive_attacks" / "seeds.txt"
+SUFFIXES_PATH = data_path / "adaptive_attacks" / "suffixes.txt"
```
Comment (jmartin-tech) [inline garak/probes/adaptive_attacks.py:28]: Prefer to align data paths to module paths.
```suggestion
SEEDS_PATH = data_path / "adaptiveattacks" / "seeds.txt"
SUFFIXES_PATH = data_path / "adaptiveattacks" / "suffixes.txt"
```

File: garak/data/adaptive_attacks/seeds.txt
Comment (jmartin-tech) [inline garak/data/adaptive_attacks/seeds.txt:1]: garak/data/adaptiveattacks/seeds.txt

File: garak/data/adaptive_attacks/suffixes.txt
Comment (jmartin-tech) [inline garak/data/adaptive_attacks/suffixes.txt:1]: ‎garak/data/adaptiveattacks/suffixes.txt

**Review by jmartin-tech — CHANGES_REQUESTED** [review]
Testing looks good, couple minor revisions.

I can make these adjustment before landing if all are in agreement.

**Comment by neerazz** [comment]
Thanks @jmartin-tech — those revisions sound good to me. Please feel free to make the adjustments before landing.

**Comment by jmartin-tech** [comment]
Minor adjustment pushed, maintainer discussion landed on naming convention as `adaptive_attacks.AdaptiveAttacks`. This will land once CI validation completes.

**Review by jmartin-tech — APPROVED** [review]
